kv/kvserver: TestNonVoterCatchesUpViaRaftSnapshotQueue failed

kv/kvserver.TestNonVoterCatchesUpViaRaftSnapshotQueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3233259&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3233259&tab=artifacts#/) on master @ [8009ac6bb63f455a18bb1139d505bdd9930816b7](https://github.com/cockroachdb/cockroach/commits/8009ac6bb63f455a18bb1139d505bdd9930816b7):


```
=== RUN   TestNonVoterCatchesUpViaRaftSnapshotQueue
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestNonVoterCatchesUpViaRaftSnapshotQueue890270849
    test_log_scope.go:74: use -show-logs to present logs inline
    replica_learner_test.go:430: 
        	Error Trace:	replica_learner_test.go:430
        	Error:      	Expect "     0.000ms      0.000ms    === operation:manual raftsnapshot queue run _unfinished:1 _verbose:1 node:1 range:41/1:/{Table/Max-Max} store:1
        	            	     0.082ms      0.082ms    event:kv/kvserver/store.go:2828 [n1,s1,r41/1:/{Table/Max-Max}] running raftsnapshot.shouldQueue
        	            	     0.107ms      0.025ms    event:kv/kvserver/store.go:2830 [n1,s1,r41/1:/{Table/Max-Max}] shouldQueue=false, priority=0.000000
        	            	" to match "streamed VIA_SNAPSHOT_QUEUE snapshot.*to.*NON_VOTER"
        	Test:       	TestNonVoterCatchesUpViaRaftSnapshotQueue
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestNonVoterCatchesUpViaRaftSnapshotQueue890270849
--- FAIL: TestNonVoterCatchesUpViaRaftSnapshotQueue (3.37s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestNonVoterCatchesUpViaRaftSnapshotQueue PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/kv aayushshah15
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestNonVoterCatchesUpViaRaftSnapshotQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -14,6 +14,7 @@ import (
 	"context"
 	"fmt"
 	"path/filepath"
+	"regexp"
 	"sort"
 	"strconv"
 	"strings"
@@ -418,16 +419,29 @@ func TestNonVoterCatchesUpViaRaftSnapshotQueue(t *testing.T) {
 	require.NotNil(t, leaseholderRepl)
 
 	time.Sleep(kvserver.RaftLogQueuePendingSnapshotGracePeriod)
-	// Manually enqueue the leaseholder replica into its store's raft snapshot
-	// queue. We expect it to pick up on the fact that the non-voter on its range
-	// needs a snapshot.
-	recording, pErr, err := leaseholderStore.ManuallyEnqueue(
-		ctx, "raftsnapshot", leaseholderRepl, false, /* skipShouldQueue */
-	)
-	require.NoError(t, pErr)
-	require.NoError(t, err)
-	trace := recording.String()
-	require.Regexp(t, "streamed VIA_SNAPSHOT_QUEUE snapshot.*to.*NON_VOTER", trace)
+
+	testutils.SucceedsSoon(t, func() error {
+		// Manually enqueue the leaseholder replica into its store's raft snapshot
+		// queue. We expect it to pick up on the fact that the non-voter on its range
+		// needs a snapshot.
+		recording, pErr, err := leaseholderStore.ManuallyEnqueue(
+			ctx, "raftsnapshot", leaseholderRepl, false, /* skipShouldQueue */
+		)
+		if pErr != nil {
+			return pErr
+		}
+		if err != nil {
+			return err
+		}
+		matched, err := regexp.MatchString("streamed VIA_SNAPSHOT_QUEUE snapshot.*to.*NON_VOTER", recording.String())
+		if err != nil {
+			return err
+		}
+		if !matched {
+			return errors.Errorf("the raft snapshot queue did not send a snapshot to the non-voter")
+		}
+		return nil
+	})
 	require.NoError(t, g.Wait())
 }
 

